### PR TITLE
Fix counter obfuscation logic

### PR DESCRIPTION
**Pull request by DataDrivenMD**
This fixes an issue that obfuscates reply, boost, and favourite counts by default on the web app. Visually, it gives users the misleading impression that there is less engagement/less activity on the Mastodon network than is actually taking place. It also creates a discordant experience for users of 3rd-party clients such as Ivory, IceCubes, and Elk-- all of which display the accurate counts.

cc @Gargron

### CODE REVIEW

**Comment by DataDrivenMD** [comment]
I just want to clarify that this PR addresses a bug, and implements the logic as it was originally intended. That is, to say, that it makes the obfuscated count an option that can be toggled on/off. As it is implemented now, the feature cannot be turned off because the logic always evaluatues to `obfuscate===true`

**Comment by trwnh** [comment]
> It also creates a discordant experience for users of 3rd-party clients such as Ivory, IceCubes, and Elk-- all of which display the accurate counts.

what 3rd-party client apps do should not have any bearing on what the official client apps do. everyone is free to build their own experience, and this experience will necessarily not be the same across every single client.

> as it was originally intended

with that said: i don't think the original intention was to make it a configurable option or toggle. i can't say that with 100% confidence, but what i *can* say is that there was an intentional design decision to not show exact counts. as initially proposed, the "reply indicator" was iconographic and not numeric *at all*. this reflected the wishes of many in the early community who did not want numbers to be shown, because showing numbers tends to lead to negative outcomes:

- people start to make decisions based on those numbers
  - a post with many favourites or boosts has a bandwagon effect. people will often favourite or boost a post simply because they see it having many favourites or boosts, not on any actual merits of the post itself. the viewer is psychologically biased and primed into seeing the post as a "good post", or at the very least a "popular post", and this creates a mental shortcut that people will take more often than not. this phenomenon is often referred to as "doing numbers".
  - a post with more replies than other interactions can lead to intentional or unintentional dogpiling, in a phenomenon often referred to as "getting ratio'd". just like viewers can be psychologically biased and primed to think a post with a lot of interactions is a "good post", they can be primed to think a post with a high reply ratio is a "bad post". again, this value judgement is made because it is easier than evaluating a post on its merits.
- there is a qualitative mental health cost associated with showing numbers
  - the mere presence of numbers immediately creates comparison, competition, external social validation, and a whole host of other effects. all of this increases the "pressure" of the "game". anecdotally (and from personal experience), having the numbers not immediately in-your-face in the timelines leads to feelings of freedom, relaxation, less stress, and so on. there is less of a perception of being "scored" or "graded".
- showing numbers encourages unwanted behaviour in general
  - the "impression of engagement and activity" is most closely associated with advertising, promotional campaigns, and commercial activity, which are all things that are particularly objectionable to the culture that has been built over the past half-decade or so. having big numbers shown to everyone is a form of "social proof" and constitutes a lot of the value behind brand deals, sponsorships, and similar engagements and agreements.
  - on an individual level, there is the drive to be popular and gain "clout". performative behaviors thrive in such an environment. aside from encouraging the worst in people, it is often the people who are best at playing these games who are also the people who should not be trusted with such status.

this is not to say that the reasoning above is some official justification, but rather, it is intended as a historical summary of discourses past. with respect to the expectations of people, please be aware and take into consideration that these expectations of people are shaped and coloured by pre-existing experiences with for-profit, corporate, ad-driven social media. their design considerations should not be our design considerations. we are not beholden to replicating their experiences. "engagement" should not be synonymous with "activity". it should be something far more qualitative.

File: app/helpers/home_helper.rb [hunk 1]
```diff
@@ -42,10 +42,8 @@ def account_link_to(account, button = '', path: nil)
   def obscured_counter(count)
     if count <= 0
       0
-    elsif count == 1
-      1
     else
-      '1+'
+      count
```
Comment (ClearlyClaire) [inline app/helpers/home_helper.rb:46]: That completely negates the purpose of `obscured_counter`

File: app/javascript/mastodon/components/animated_number.js [hunk 2]
```diff
@@ -51,7 +51,7 @@ export default class AnimatedNumber extends React.PureComponent {
     const { direction } = this.state;
 
     if (reduceMotion) {
-      return obfuscate ? obfuscatedCount(value) : <ShortNumber value={value} />;
+      return (obfuscate === true) ? obfuscatedCount(value) : <ShortNumber value={value} />;
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/components/animated_number.js:54]: I am not sure what this attempts to fix.

File: app/javascript/mastodon/components/animated_number.js [hunk 3]
```diff
@@ -65,7 +65,7 @@ export default class AnimatedNumber extends React.PureComponent {
         {items => (
           <span className='animated-number'>
             {items.map(({ key, data, style }) => (
-              <span key={key} style={{ position: (direction * style.y) > 0 ? 'absolute' : 'static', transform: `translateY(${style.y * 100}%)` }}>{obfuscate ? obfuscatedCount(data) : <ShortNumber value={data} />}</span>
+              <span key={key} style={{ position: (direction * style.y) > 0 ? 'absolute' : 'static', transform: `translateY(${style.y * 100}%)` }}>{(obfuscate === true) ? obfuscatedCount(data) : <ShortNumber value={data} />}</span>
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/components/animated_number.js:68]: I'm not sure what this attempts to fix.

File: app/javascript/mastodon/components/status_action_bar.js [hunk 4]
```diff
@@ -360,8 +360,8 @@ class StatusActionBar extends ImmutablePureComponent {
...
-        <IconButton className={classNames('status__action-bar__button', { reblogPrivate })} disabled={!publicStatus && !reblogPrivate} active={status.get('reblogged')} title={reblogTitle} icon='retweet' onClick={this.handleReblogClick} counter={withCounters ? status.get('reblogs_count') : undefined} />
-        <IconButton className='status__action-bar__button star-icon' animate active={status.get('favourited')} title={intl.formatMessage(messages.favourite)} icon='star' onClick={this.handleFavouriteClick} counter={withCounters ? status.get('favourites_count') : undefined} />
+        <IconButton className={classNames('status__action-bar__button', { reblogPrivate })} disabled={!publicStatus && !reblogPrivate} active={status.get('reblogged')} title={reblogTitle} icon='retweet' onClick={this.handleReblogClick} counter={(withCounters !== false) ? status.get('reblogs_count') : undefined} />
+        <IconButton className='status__action-bar__button star-icon' animate active={status.get('favourited')} title={intl.formatMessage(messages.favourite)} icon='star' onClick={this.handleFavouriteClick} counter={(withCounters !== false) ? status.get('reblogs_count') : undefined} />
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/components/status_action_bar.js:364]: This is a really convoluted way to switch the default, and therefore negate the use of `withCounters` as it's only ever used to enable the counts (on the “Explore” tab), never to explicitly disable them (which is the intended default)

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
>  just want to clarify that this PR addresses a bug, and implements the logic as it was originally intended. That is, to say, that it makes the obfuscated count an option that can be toggled on/off. As it is implemented now, the feature cannot be turned off because the logic always evaluatues to `obfuscate===true`

I'm afraid that simply isn't true, the counts are purposefully hidden and obfuscated everywhere besides the detailed post view and the “Explore” tab, and those props are not user-configurable but used as flags in the code.

Maybe it could be changed to be an user option, but the original intent *was* to hide those numbers in timelines, to avoid focusing on the popularity of content and distorted incentives as @trwnh pointed out. These design choices are commonly lauded, but are not universally agreed upon either.

**Comment by ClearlyClaire** [comment]
Reply counts are always obfuscated, but other counts are not, so the prop in the `AnimatedNumber` component (which is used for all kinds of counts) is justified.

`StatusActionBar` used to not never display counts for favs and boosts and thus not have the `withCounters` prop, but since the inclusion of the explore tab, which does display (a design decision I'm not a fan of) those counts using the same component. So I feel the prop is justified too.

The detailed view also displays those counts, and does not require another API request.

File: app/javascript/mastodon/components/animated_number.js (hunk 2 above)
Comment (DataDrivenMD) [reply app/javascript/mastodon/components/animated_number.js:54]: AFAICT the value of `obfuscate` is not initialized when this component is loaded, so the conditional logic in the original code doesn't actually do anything. Unless I'm missing upstream initialization the the conditional as originally coded will always evaluate to `false`. If that's the intent, then the conditional is entirely unnecessary.

File: app/javascript/mastodon/components/animated_number.js (hunk 3 above)
Comment (DataDrivenMD) [reply app/javascript/mastodon/components/animated_number.js:68]: Same as above

File: app/helpers/home_helper.rb (hunk 1 above)
Comment (DataDrivenMD) [reply app/helpers/home_helper.rb:46]: Sorry, this was meant to be
```
def obscured_counter(count)
    if count <= 0
      0
    elsif count == 1
      1
    else
      count
```

Comment (DataDrivenMD) [reply app/helpers/home_helper.rb:46]: Wait, actually, upon further review, I didn't mean for this change to be there. At all.

**Comment by DataDrivenMD** [comment]
Sorry, I'm having trouble tracking the shifting justifications for inaction, here. When I first mentioned it on Mastodon, @Gargron made it sound like the obfuscated counts were not functioning according to spec. It wasn't until **after** I volunteered a PR that he said it was intentional, and added that he didn't ["feel very strongly in favour of it"](https://mastodon.social/@Gargron/109793474325459708).

Here, the PR review began with vague assertions based on vague recollections of the original intent and line-level comments that suggest a cursory PR review, at best. Now, after I pointed out that even when the original intent is factored into the equation, there are logical flaws in the code, the rationale has shifted to personal preference:
>   but since the inclusion of the explore tab, which does display (a design decision I'm not a fan of) those counts using the same component. So I feel the prop is justified too.

I honestly don't care what you do with this PR. I was simply trying to help. I just wish that I hadn't been led to believe that contributing to this project was a productive use of my time. Next time, I would start the PR review with: "Outside contributions are not welcome unless they've been pre-approved by [X, Y, and/or Z]"

File: app/javascript/mastodon/components/status_action_bar.js (hunk 4 above)
Comment (ClearlyClaire) [reply app/javascript/mastodon/components/status_action_bar.js]: > This is another example where the variable is not initialized, so the conditional logic doesn't actually do anything. In this case, the `withCounters` variable is not initialized, so the conditional logic will always evaluate to `false`. This line must be paired with the change I made in status_list.js to yield a working conditional.

`withCounters` is passed from `StatusList`, and `StatusList` has its `withCounters` set to `true` in `app/javascript/mastodon/features/explore/statuses.js`. This is as explained: it is set in the Explore tab only.

> None of these conditionals rely on the application's context/global state. None check the `window` element to see where the user happens to be within the application's navigation tree. If there's some contextual logic that stems from the patchwork of undefined variables, then that logic should be made explicit.

We're not checking on the `window` element, we're passing the prop down where it matters.

**Comment by ClearlyClaire** [comment]
> Sorry, I'm having trouble tracking the shifting justifications for inaction, here. When I first mentioned it on Mastodon, @Gargron made it sound like the obfuscated counts were not functioning according to spec. It wasn't until **after** I volunteered a PR that he said it was intentional, and added that he didn't ["feel very strongly in favour of it"](https://mastodon.social/@Gargron/109793474325459708).

Your original question was about under-reporting numbers but the PR is about where to show the numbers. I think it's not a shift in justification, it's just two different (though related) things. The first could have been a bug, the second is a design decision some people disagree with.

**Comment by Gargron** [comment]
> Sorry, I'm having trouble tracking the shifting justifications for inaction, here. When I first mentioned it on Mastodon, @Gargron made it sound like the obfuscated counts were not functioning according to spec.

You didn't mention that you were talking about obfuscated counts. You made it sound like there was a bug that made the *numbers* lower than they should be. Let's be clear about that.

**Pull request closed without merging**